### ISSUE TITLE
Roadmap for addressing scaling? How about an internal scale service?

### ISSUE THREAD

**Issue by chrisstpierre** [storyscript/platform-engine#366]
@wilzbach my worry about this issue is that we are essentially taking random data points and hoping they are an accurate description of resource use. We are going to have a lot of services that are mostly unused, with occasional spikes in resource use, & we need enough resources supplied to be able to manage those spikes. IE: ImageMagick might be used a few times per day, but when it is used it requires a ton of resources, and our data collection this way doesn't help us prepare for that really. Without enough resources ImageMagick will just crash. @anukul and I had a discussion about this as well.

RE: dynamically down/up-size pods - We definitely want to do this - but hold on because it's not that easy. Of course we could scale up when use increases, but that might not actually help anything. What we would really need for it to be useful is a machine learning service hooked up to our Prometheus data. Predictive Scaling: [aws predictive scaling](https://aws.amazon.com/blogs/aws/new-predictive-scaling-for-ec2-powered-by-machine-learning/) [ai kubernetes autoscaling](https://docs.magalix.com/docs/how-magalix-ai-powers-applications). If we don't do that then we are behind the curve. 

I would also suggest giving advanced users the ability to choose when their services scale. I'm not sure how it would look, but maybe in the story?
?
```
when http server listen method:'post' path:'/upload' as req
    dest = '/uploads/{awesome id}.png'
    image = req.files['image']
    image_bytes = image['size']

    needed_memory = round (image_bytes/1048576)*2
    needed_cpu = round needed_memory/3

    scale imagemagick mem:"{needed_memory}M" cpu:"{needed_cpu}m"
    pdf= imagemagik convert image:image['binary'] to:pdf width:2000 height:5000

    aws/s3 put content:img :dest
    req write content:{location:dest}
    req finish
```

Because for this type of scenario no predictive scaling can help. So what else could we do besides this?
We couldn't let the service scale itself either, because the service doesn't know the business case. If we need this image converted FAST then we need to give more resources. If we are okay with slower processing then we can choose something smaller to save money.

And then I suppose the service should go back to its default size after the line of last occurrence is finished executing. 

Ok this is a long comment I will make a new issue for this.

_Originally posted by @chrisstpierre in https://github.com/storyscript/platform-engine/pull/346#issuecomment-508294634_

**Comment by wilzbach** [comment]
@chrisstpierre: it was just a point that I thought was very related to that PR.

> and our data collection this way doesn't help us prepare for that really. Without enough resources ImageMagick will just crash. 

Well, one point that I made in that PR is that if we plan to push the data into Prometheus anyhow, why not directly do it in that PR.

> What we would really need for it to be useful is a machine learning service hooked up to our Prometheus data.

Predictive scaling doesn't work too well for a lot of use cases because you can't always predict requests. That's why a lot of people still prefer the conservative scaling algorithms as for them latency and availability is more important than saving a few bucks.
Anyhow, while AWS does support up-scaling, vertical scaling with the same machine type is more popular and it's fine if we don't want to support automatic service up-scaling either. It was just something that should be considered.

> I would also suggest giving advanced users the ability to choose when their services scale. I'm not sure how it would look, but maybe in the story?

I don't think this will work. Doing this via the language looks a bit ugly and should be the last resort. There are a few other things that we can explore in the meantime:

- advanced users can set 'minimal memory' in their microservice yaml
- if we detect that a service uses a ton of resources when working, but almost none when asleep, which is 99% of the time, we can downscale/suspend that service and only allocate more resources before a request. This goes into the heroku suspend model which we want to support too, but I guess this is a bit of different discussion as we might be able to use docker suspend here instead of terminating instances.

**Comment by chrisstpierre** [comment]
> Well, one point that I made in that PR is that if we plan to push the data into Prometheus anyhow, why not directly do it in that PR.
I agree. I think its a good time to discuss our medium-term approach.


> Predictive scaling doesn't work too well for a lot of use cases because you can't always predict requests. That's why a lot of people still prefer the conservative scaling algorithms as for them latency and availability is more important than saving a few bucks.
> Anyhow, while AWS does support up-scaling, vertical scaling with the same machine type is more popular and it's fine if we don't want to support automatic service up-scaling either. It was just something that should be considered.

Although when we do get to the point when we have tons of data, we can do a pretty good job.  But that's really down the road.

> I don't think this will work. Doing this via the language looks a bit ugly and should be the last resort. There are a few other things that we can explore in the meantime:
> 
> * advanced users can set 'minimal memory' in their microservice yaml
> * if we detect that a service uses a ton of resources when working, but almost none when asleep, which is 99% of the time, we can downscale/suspend that service and only allocate more resources before a request. This goes into the heroku suspend model which we want to support too, but I guess this is a bit of different discussion as we might be able to use docker suspend here instead of terminating instances.

I agree that these two strategies combined will solve 95% of the more challenging use cases. If its being used like an ETL then maybe there's a message queue keeping things stable. Otherwise let's say someone wants to put in a few gigabytes of data into a service that processes it in-memory, they could just set the minimum to be extra high as long as it does scale back down again afterwards. OR maybe the services can control their own resources somehow.

min memory in the yaml makes sense. We'll need to think more about docker suspend 🤷‍♂
